compositor_render/src/transformations/web_renderer/browser_client.rs

```diff
@@ -81,6 +53,61 @@ impl BrowserClient {
             frame_data,
             source_transforms,
             resolution,
+            chromium_thread_sender: None,
+        }
+    }
+
+    pub fn set_chromium_thread_sender(&mut self, sender: Sender<ChromiumSenderMessage>) {
+        self.chromium_thread_sender.replace(sender);
+    }
+
+    fn update_source_transforms(&self, msg: &cef::ProcessMessage) {
+        let mut transforms_matrices = Vec::new();
+        for i in (0..msg.size()).step_by(4) {
+            let Some(x) = msg.read_double(i) else {
+                error!("Expected \"x\" value of DOMRect");
+                continue;
+            };
+            let Some(y) = msg.read_double(i + 1) else {
+                error!("Expected \"y\" value of DOMRect");
+                continue;
+            };
+            let Some(width) = msg.read_double(i + 2) else {
+                error!("Expected \"width\" value of DOMRect");
+                continue;
+            };
+            let Some(height) = msg.read_double(i + 3) else {
+                error!("Expected \"height\" value of DOMRect");
+                continue;
+            };
+
+            let transformations_matrix = BoxLayout {
+                top_left_corner: (x as f32, y as f32),
+                width: width as f32,
+                height: height as f32,
+                rotation_degrees: 0.0,
+            }
+            .transformation_matrix(self.resolution);
+
+            transforms_matrices.push(transformations_matrix);
+        }
+
+        let mut source_transforms = self.source_transforms.lock().unwrap();
+        *source_transforms = transforms_matrices;
+    }
+
+    fn handle_unembed_source_confirmation(&self, msg: &cef::ProcessMessage) {
+        let Some(shared_memory_path) = msg.read_string(0) else {
+            error!("Expected shared memory path");
+            return;
+        };
+        debug!("Process helper successfuly unembedded source. \"{shared_memory_path}\" is no longer used");
+        if let Some(chromium_thread_sender) = &self.chromium_thread_sender {
+            chromium_thread_sender
+                .send(ChromiumSenderMessage::FinalizePendingResize {
+                    shared_memory_path: shared_memory_path.into(),
+                })
+                .unwrap()
```

Comment:
is this safe?

Reply:
I'm not 100% sure. I will add error checking here.